### PR TITLE
Add upgrade check for dropped columns of different length

### PR DESCRIPTION
**Pull request by dgkimura**
This PR adds additional check for non-upgradable cases involving partitioned tables with dropped columns.  It includes a refactor of the integration testing for non-passing check tests.  A caveat is that we no longer use "then" function which is tightly coupled with happy path of upgrade that starts up the GPDB 6 cluster.

### CODE REVIEW

File: contrib/pg_upgrade/test/integration/scenarios/heterogeneous_partitioned_heap_table.c [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	when(anAdministratorPerformsAnUpgradeCheck);
+	thenCheckFailsWithError("Your installation contains heterogenous partitioned tables where the root partition does not match one or more child partition on disk representation.");
+}
+
+void
+test_a_partitioned_heap_table_with_differently_sized_dropped_columns_cannot_be_upgraded(void **state)
+{
+	given(createPartitionedHeapTableWithDifferntlySizedDroppedColumsInFiveCluster);
+	when(anAdministratorPerformsAnUpgradeCheck);
+	thenCheckFailsWithError("Your installation contains heterogenous partitioned tables where the root partition does not match one or more child partition on disk representation.");
```
Comment (berlin-ab) [inline contrib/pg_upgrade/test/integration/scenarios/heterogeneous_partitioned_heap_table.c]: In principle, the scenario outline should not include the specifics of the error message - to keep the level of abstraction the same in each test function we don't want to mix high-level scenario with low-level specifics.

File: contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.h [hunk 2]
```diff
@@ -2,7 +2,13 @@
...
 
+#include "postgres_fe.h"
+#include "pqexpbuffer.h"
+
 void		performUpgrade(void);
-void		performUpgradeCheckFailsWithError(char *message);
+void		performUpgradeCheck();
+
+extern PQExpBufferData pg_upgrade_output;
+extern int pg_upgrade_exit_status;
```
Comment (berlin-ab) [inline contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.h]: Can we hide these state variables behind an API? 

`upgradeCheckStatus()` for example?

File: contrib/pg_upgrade/test/integration/utilities/test-helpers.h [hunk 3]
```diff
@@ -2,7 +2,8 @@
 
 PGconn *connectToFive(void);
 void resetGpdbFiveDataDirectories(void);
+void initializePgUpgradStatus(void);
 
 PGconn *connectToSix(void);
 void resetGpdbSixDataDirectories(void);
-
+void resetPgUpgradeStatus(void);
```
Comment (berlin-ab) [inline contrib/pg_upgrade/test/integration/utilities/test-helpers.h]: Seems like this goes with the `upgrade-helpers`

File: contrib/pg_upgrade/test/integration/utilities/test-helpers.c [hunk 4]
```diff
@@ -3,6 +3,21 @@
 
 #include "test-helpers.h"
 #include "query-helpers.h"
+#include "upgrade-helpers.h"
+#include "pqexpbuffer.h"
+
+void
+initializePgUpgradStatus(void)
```
Comment (berlin-ab) [inline contrib/pg_upgrade/test/integration/utilities/test-helpers.c]: s/PgUpgrad/PgUpgrade/

File: contrib/pg_upgrade/test/integration/scenarios/heterogeneous_partitioned_heap_table.c (hunk 1 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/scenarios/heterogeneous_partitioned_heap_table.c]: Good to know. I pushed the error message into the function renamed to `thenCheckFailsHeterogeneousPartitionedTableCheck()`

File: contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.h (hunk 2 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.h]: Sounds good.  Created 2 functions to do that: 
```c
char *upgradeCheckOutput(void);
int upgradeCheckStatus(void);
```

File: contrib/pg_upgrade/test/integration/utilities/test-helpers.h (hunk 3 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/utilities/test-helpers.h]: Yup, that works better.  Then no need to export global variables for `upgradeCheck*` functions.

File: contrib/pg_upgrade/test/integration/utilities/test-helpers.c (hunk 4 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/utilities/test-helpers.c]: Done, thanks.

File: contrib/pg_upgrade/check_gp.c [hunk 5]
```diff
@@ -366,6 +368,134 @@ check_orphaned_toastrels(void)
 
 }
 
+static void
+check_heterogeneous_partition(void)
```
Comment (berlin-ab) [inline contrib/pg_upgrade/check_gp.c]: I like that we have coverage for many different scenarios for this function, although I would like us to think about how we could add lower level tests for these checks. Most of the logic for `check_heterogeneous_partition` is the same for the variations. The variations really come out of the complex query that we've written. If we could isolate the query into its own function, then we could test that the query returns us rows for the scenarios that we're interested in - while also having a test that says "if there are rows returned from the query, then fail"

Comment (berlin-ab) [reply contrib/pg_upgrade/check_gp.c]: We could leave one fully-integrated test to make sure everything is still hooked up correctly.

File: contrib/pg_upgrade/check_gp.c [hunk 6]
```diff
@@ -366,6 +368,134 @@ check_orphaned_toastrels(void)
...
+			for (rowno = 0; rowno < ntups; rowno++)
+				fprintf(script, "  %s has different number of columns in a child and root partition\n",
+						PQgetvalue(res, rowno, PQfnumber(res, "parrelid")));
+
+			fclose(script);
+		}
+
+		PQclear(res);
+
+		res = executeQueryOrDie(conn,
```
Comment (berlin-ab) [inline contrib/pg_upgrade/check_gp.c]: are we really doing two checks in one here?

File: contrib/pg_upgrade/check_gp.c [hunk 7]
```diff
@@ -366,6 +368,134 @@ check_orphaned_toastrels(void)
...
+		PQfinish(conn);
+
+	}
+
+	if (found)
+	{
+		pg_log(PG_REPORT, "fatal\n");
+		pg_log(PG_FATAL,
+			   "| Your installation contains heterogenous partitioned tables where the root partition does not match one or more child partition on disk representation.\n"
+			   "| In order to make the tables homogeneous, CTAS must be performed on this table so that it can be be dropped before upgrade.\n"
```
Comment (berlin-ab) [inline contrib/pg_upgrade/check_gp.c]: Does the end-user know what CTAS means?

File: contrib/pg_upgrade/check_gp.c (hunk 5 above)
Comment (dgkimura) [reply contrib/pg_upgrade/check_gp.c]: Seems like a good idea. I'll give this some thought.

File: contrib/pg_upgrade/check_gp.c (hunk 7 above)
Comment (dgkimura) [reply contrib/pg_upgrade/check_gp.c]: Good point.  That acronym is never used in the docs. Will change.

File: contrib/pg_upgrade/check_gp.c (hunk 6 above)
Comment (dgkimura) [reply contrib/pg_upgrade/check_gp.c]: Yeah, though I think they're related to the same class of error.  Either number of columns is off or attributes of the columns are off.  `check_is_super_user` is a similar example that can fail in different ways for the same class of error.

File: contrib/pg_upgrade/test/check_gp_test.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+	expect_value(PQclear, res, result_addr);
+	will_be_called(PQclear);
+
+	/*
+	 * And the query to check different columns attributes returns 0 results
+	 */
+	expect_any(executeQueryOrDie, conn);
+	expect_string(executeQueryOrDie, fmt,
+		"SELECT parrelid::regclass, att1.attnum, rule.parchildrelid::regclass, att1.attname attname1, att2.attname attname2, att1.attisdropped attisdropped1, att2.attisdropped attisdropped2, att1.attlen attlen1, att2.attlen attlen2, att1.atttypid atttypid1, att2.atttypid atttypid2, att1.attalign attalign1, att2.attalign attalign2 "
```
Comment (berlin-ab) [inline contrib/pg_upgrade/test/check_gp_test.c]: The implementation detail here is a signal we could improve something. Maybe a constant to hold this query, that we can reference here and in the prod code? We don't want to update two places if we make a change to this query.

Comment (dgkimura) [reply contrib/pg_upgrade/test/check_gp_test.c]: Done

**Comment by gkokolatos** [comment]
May I ask which is the status of this @dgkimura ? I think it looks good and the comments are addressed.

**Comment by dgkimura** [comment]
On Tue, Nov 5, 2019 at 12:41 AM gkokolatos wrote:
>
> May I ask which is the status of this @dgkimura ? I think it looks good and the comments are addressed.

Thanks for asking.  It's ready to go, just didn't see explicit +1.  I'll merge today

**Pull request merged**